artifact/sha/sha.cpp

```diff
@@ -0,0 +1,126 @@
+#include <artifact/sha/sha.hpp>
+
+#include <common/common.hpp>
+
+#include <cstdint>
+#include <string>
+#include <ctime>
+#include <iomanip>
+#include <vector>
+#include <sstream>
+#include <algorithm>
+
+namespace mender {
+namespace sha {
+
+namespace log = mender::common::log;
+
+
+const ErrorCategoryClass ErrorCategory = ErrorCategoryClass();
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "ShaSumErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	InitializationError:
+		return "Initialization error";
+	ShasumCreationError:
+		return "Shasum creation error";
+	ShasumMismatchError:
+		return "Shasum mismatch error";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code) {
+	return error::Error(error_condition(code, ErrorCategory), "");
+}
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+Reader::Reader(io::Reader &reader) :
+	Reader::Reader {reader, ""} {
+}
+
+Reader::Reader(io::Reader &reader, std::string expected_sha = "") :
+	sha_handle_(EVP_MD_CTX_new(), [](EVP_MD_CTX *ctx) { EVP_MD_CTX_free(ctx); }),
+	wrapped_reader_ {reader},
+	buffer_(1024),
+	expected_sha_ {expected_sha} {
+	if (EVP_DigestInit_ex(sha_handle_.get(), EVP_sha256(), nullptr) != 1) {
+		log::Error("Failed to initialize the shasummer");
+		initialized = false;
+		return;
+	}
+	initialized = true;
+}
+
+expected::ExpectedSize Reader::Read(vector<uint8_t> &dst) {
+	if (!initialized) {
+		return MakeError(InitializationError);
+	}
+
+	auto bytes_read = wrapped_reader_.Read(buffer_);
```

Comment:
This is incorrect. The buffer needs to be resized to `min(dst.size(), buffer_.size())`. Otherwise it will start losing bytes because it only puts `dst.size()` bytes into that buffer and anything overshooting that, is lost. Then it needs to resized back to original capacity afterwards.

In general I have wondered if using direct `vector`s was a mistake, and we should have had this interface instead:

```
ExpectedSize Read(vector<uint8_t>::iterator start, vector<uint8_t>::iterator end);
```

This aligns better with how the rest of library works, and is also closer, at least in philosophy, to the Golang slices. Do you think we should consider mass-changing it?

Reply:
hah, I just noticed it actually :p

Reply:
Yes, I was just thinking of writing something similar actually, cause I've been runnning into this very same problem in the `tar-reader` implementation too. It's really annoying handling cases in which `size(dst) < size(buf)`, and you have to do intermediate caching, returning, and then fill your buffer anew.

Reply:
I had this beautiful `// TODO - What if 'size(dst) < size(buf)'` in my notes :sweat:

Reply:
Just to conclude: Ole will fix the function using the current interface, and I will take a look at changing our interface to use iterators instead. But this will be in a followup.

Reply:
For now, I solved it by adding a buf-resizer class, which resizes on destruction.

Reply:
See: https://github.com/mendersoftware/mender/pull/1160/commits/365700938f8bc1da15a5d60f402ceb21d951b0ac#diff-2c09a25897ac9f57ed5e29eaec90863a06d6656551268cffa1c8497f92291166R72